internal/webhook/external/azure.go

```diff
@@ -0,0 +1,222 @@
+	azure              = "azure"
+
+	acrPingEvent         = "ping"
+	acrPushEvent         = "push"
+	azureDevOpsPushEvent = "git.push"
+
+	acrUserAgentPrefix         = "AzureContainerRegistry"
+	azureDevOpsUserAgentPrefix = "VSServices"
+)
+
+func init() {
+	registry.register(
+		azure,
+		webhookReceiverRegistration{
+			predicate: func(cfg kargoapi.WebhookReceiverConfig) bool {
+				return cfg.Azure != nil
+			},
+			factory: newAzureWebhookReceiver,
+		},
+	)
+}
+
+// azureWebhookReceiver is an implementation of WebhookReceiver that handles
+// inbound webhooks from both Azure Container Registry (ACR) and Azure DevOps.
+type azureWebhookReceiver struct {
+	*baseWebhookReceiver
+}
+
+// newAzureWebhookReceiver returns a new instance of azureWebhookReceiver.
+func newAzureWebhookReceiver(
+	c client.Client,
+	project string,
+	cfg kargoapi.WebhookReceiverConfig,
+) WebhookReceiver {
+	return &azureWebhookReceiver{
+		baseWebhookReceiver: &baseWebhookReceiver{
+			client:     c,
+			project:    project,
+			secretName: cfg.Azure.SecretRef.Name,
+		},
+	}
+}
+
+// getReceiverType implements WebhookReceiver.
+func (a *azureWebhookReceiver) getReceiverType() string {
+	return azure
+}
+
+// getSecretValues implements WebhookReceiver.
+func (a *azureWebhookReceiver) getSecretValues(
+	secretData map[string][]byte,
+) ([]string, error) {
+	secretValue, ok := secretData[azureSecretDataKey]
+	if !ok {
+		return nil,
+			fmt.Errorf("missing data key %q for Azure WebhookReceiver", azureSecretDataKey)
+	}
+	return []string{string(secretValue)}, nil
+}
+
+// getHandler implements WebhookReceiver.
+func (a *azureWebhookReceiver) getHandler(requestBody []byte) http.HandlerFunc {
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		ctx := r.Context()
+		logger := logging.LoggerFromContext(ctx)
+		logger.Debug("received azure event", "userAgent", r.UserAgent())
+
+		switch {
+		// Format is AzureContainerRegistry/<version>
+		case strings.HasPrefix(r.UserAgent(), acrUserAgentPrefix):
+			a.handleACREvent(ctx, w, requestBody)
+		// Format is VSServices/<version>
+		case strings.HasPrefix(r.UserAgent(), azureDevOpsUserAgentPrefix):
+			a.handleAzureDevOpsEvent(ctx, w, requestBody)
+		default:
+			xhttp.WriteErrorJSON(
+				w,
+				xhttp.Error(
+					errors.New(
+						"request does not appear to have originated from a supported service",
+					),
+					http.StatusBadRequest,
+				),
+			)
+			return
+		}
+	})
+}
+
+func (a *azureWebhookReceiver) handleACREvent(
+	ctx context.Context,
+	w http.ResponseWriter,
+	requestBody []byte,
+) {
+	var event acrEvent
+	if err := json.Unmarshal(requestBody, &event); err != nil {
+		xhttp.WriteErrorJSON(
+			w,
+			xhttp.Error(errors.New("invalid request body"),
+				http.StatusBadRequest,
+			),
+		)
+		return
+	}
+
+	switch event.Action {
+	case acrPushEvent:
+		repoURL := normalizeOCIRepoURL(
+			fmt.Sprintf("%s/%s", event.Request.Host, event.Target.Repository),
+			event.Target.MediaType,
+		)
```

Comment:
Is there media type info in the payload? We should try to determine if the push was an image or a chart because their URLs use different normalization processes.

Reply:
Nevermind. This is using the new function already. I'm blind.